arrows/vxl/convert_image.cxx

```diff
@@ -0,0 +1,498 @@
+template < typename InputType, typename OutputType >
+void
+percentile_scale_image(
+  vil_image_view< InputType > const& src, vil_image_view< OutputType >& dst,
+  double lower, double upper, unsigned sampling_points,
+  bool ignore_extremes = true )
+{
+  std::vector< double > percentiles( 2, 0.0 );
+  percentiles[ 0 ] = lower;
+  percentiles[ 1 ] = upper;
+
+  std::vector< InputType > percentile_values =
+    get_image_percentiles( src, percentiles, sampling_points,
+                           ignore_extremes );
+
+  OutputType max_val = std::numeric_limits< OutputType >::max();
+
+  InputType lower_bound = percentile_values[ 0 ];
+  InputType upper_bound = percentile_values[ 1 ];
+
+  double scale;
+
+  if( percentile_values[ 1 ] - percentile_values[ 0 ] > 0 )
+  {
+    scale = ( static_cast< double >( max_val ) + 0.5 ) /
+            ( percentile_values[ 1 ] - percentile_values[ 0 ] );
+  }
+  else
+  {
+    scale = static_cast< double >( max_val ) /
+            std::numeric_limits< InputType >::max();
+  }
+
+  auto const ni = src.ni();
+  auto const nj = src.nj();
+  auto const np = src.nplanes();
+  dst.set_size( ni, nj, np );
+
+  // Stretch image to upper and lower percentile bounds
+  vil_transform( src, dst,
+                 [ lower_bound, upper_bound, scale ]( InputType pixel ){
+                   if( pixel < lower_bound )
+                   {
+                     return static_cast< OutputType >( 0 );
+                   }
+                   else if( pixel > upper_bound )
+                   {
+                     return static_cast< OutputType >( std::numeric_limits< OutputType >
+                                                       ::max() );
+                   }
+                   else
+                   {
+                     return static_cast< OutputType >( ( pixel -
+                                                         lower_bound ) * scale );
+                   }
+                 } );
+}
+
+} // namespace <anonoymous>
+
+// ----------------------------------------------------------------------------
+/// Private implementation class
+class convert_image::priv
+{
+public:
+  priv()
+    : format{ "byte" }
+      , single_channel{ false }
+      , scale_factor{ 0.0 }
+      , random_greyscale{ 0.0 }
+      , percentile_norm{ -1.0 }
+  {
+  }
+
+  // Convert the type
+  template < typename ipix_t > vil_image_view< ipix_t >
+  convert( vil_image_view_base_sptr& view );
+
+  // Scale and convert the image
+  template < typename opix_t, typename ipix_t > vil_image_view< opix_t >
+  scale( vil_image_view< ipix_t > input );
+
+  std::string format;
+  bool single_channel;
+  double scale_factor;
+  double random_greyscale;
+  double percentile_norm;
+};
+
+// ----------------------------------------------------------------------------
+template < typename ipix_t >
+vil_image_view< ipix_t >
+convert_image::priv
+::convert( vil_image_view_base_sptr& view )
```

Comment:
Is this called from anywhere?

Reply:
If not, then the functions (`random_grey_conversion`, and `combine_channels`) and config options (`single_channel`) used below are also not used anywhere else that I can see.

Reply:
Yes, in line 454 in this file. I've updated it to be named in a more informative and distinct way.